src/audit_payloads.py

```diff
@@ -0,0 +1,194 @@
+
+maxjobsecs = None
+runningjobs = {}
+firstidx = None
+lastidx = None
+
+def dumpvars():
+    htcondor.log(htcondor.LogLevel.Audit, "firstidx: %s, lastidx: %s, runningjobs: %s" % (firstidx, lastidx, str(runningjobs)))
+
+# stop tracking a running job
+def removerunningjob(idx):
+    global runningjobs
+    global firstidx
+    global lastidx
+
+    if idx not in runningjobs:
+	# shouldn't happen, but just in case
+	return
+
+    # remove this job from the doubly-linked list
+    thisjob = runningjobs[idx]
+    if thisjob['previdx'] == None:
+	firstidx = thisjob['nextidx']
+    else:
+	runningjobs[thisjob['previdx']]['nextidx'] = thisjob['nextidx']
+    if thisjob['nextidx'] == None:
+	lastidx = thisjob['nextidx']
+    else:
+	runningjobs[thisjob['nextidx']]['previdx'] = thisjob['previdx']
+
+    # and remove it from the running jobs
+    del runningjobs[idx]
+
+
+# start tracking a running job
+def addrunningjob(idx, globaljobid, now):
+    global runningjobs
+    global firstidx
+    global lastidx
+
+    if idx in runningjobs:
+	# shouldn't happen, but just in case
+	removerunningjob(idx)
+
+    thisjob = {}
+    thisjob['globaljobid'] = globaljobid
+    thisjob['starttime'] = now
+
+    # append this job the end of the doubly-linked list
+    thisjob['previdx'] =  lastidx
+    thisjob['nextidx'] =  None
+    if lastidx != None:
+	runningjobs[lastidx]['nextidx'] = idx
+    lastidx = idx
+    if firstidx == None:
+	firstidx = idx 
+
+    # and add it to the running jobs
+    runningjobs[idx] = thisjob
+
+# a job may be being stopped
+def stopjob(info):
+    global runningjobs
+    if 'Name' not in info or 'SlotID' not in info:
+	return
+    idx = (info['Name'], info['SlotID'])
+    if idx not in runningjobs:
+	return
+    loginfo = {}
+    loginfo['Name'] = info['Name']
+    loginfo['SlotID'] = info['SlotID']
+    loginfo['GlobalJobId'] = runningjobs[idx]['GlobalJobId']
+    htcondor.log(htcondor.LogLevel.Audit, "Job stop: %s" % loginfo)
+
+    removerunningjob(idx)
+    #dumpvars()
+
+# a job may be being started
+def startjob(info):
+    global maxjobsecs
+
+    if 'Name' not in info or 'SlotID' not in info or 'GlobalJobId' not in info:
+	return
+
+    idx = (info['Name'], info['SlotID'])
+    globaljobid = info['GlobalJobId']
+    if idx in runningjobs:
+	if globaljobid == runningjobs[idx]['globaljobid']:
+	    # just an update to a running job, ignore
+	    return
+	# first stop the existing job, the slot is being reused
+	stopjob(info)
+
+    htcondor.log(htcondor.LogLevel.Audit, "Job start: %s" % info)
+    now = datetime.now()
+    addrunningjob(idx, globaljobid, now)
+
+    if maxjobsecs == None:
+	# initialize this global variable
+	if 'AUDIT_PAYLOAD_MAX_HOURS' in htcondor.param:
+	    maxjobhours = int(htcondor.param['AUDIT_PAYLOAD_MAX_HOURS'])
+	else:
+	    maxjobhours = 3 * 24
+	htcondor.log(htcondor.LogLevel.Audit,
+	    "Audit payload maximum job hours: %d" % maxjobhours)
+	maxjobsecs = maxjobhours * 60 * 60
```

Comment:
Rather than caching it on first call, can you simply do this at the module level when things are loaded?